File: python/src/patterns/creational/builder/query.py
class Query:
    # required
    _columns: list[str]
    _table: str
    # optional
    _where_clauses: list[str] | None
    _order_by: str | None
    _limit: int | None
    _offset: int | None
    # can add more fields as needed

    def __init__(self, builder: "Builder") -> None:
        self._columns = builder._columns
        self._table = builder._table
        self._where_clauses = builder._where_clauses
        self._order_by = builder._order_by
        self._limit = builder._limit
        self._offset = builder._offset

    def __repr__(self) -> str:
        return f"Query(table={self._table}, columns={self._columns}, where_clauses={self._where_clauses}, order_by={self._order_by}, limit={self._limit}, offset={self._offset})"

    def to_sql(self) -> str:
        if len(self._columns) == 1 and self._columns[0] == "*":
            columns = "*"
        else:
            columns = ", ".join(self._columns)
        sql = f"SELECT {columns} FROM {self._table}"
        if self._where_clauses:
            sql += f" WHERE {' AND '.join(self._where_clauses)}"
        if self._order_by:
            sql += f" ORDER BY {self._order_by}"
        if self._limit:
            sql += f" LIMIT {self._limit}"
        if self._offset:
            sql += f" OFFSET {self._offset}"
        return sql + ";"

    class Builder:
        def __init__(self, table: str, *args) -> None:
            self._table = table
            self._columns = list(args) if args else ["*"]
            self._where_clauses = []
            self._order_by = None
            self._limit = None
            self._offset = None

        # for each method, we return the Builder object itself
        def where(self, condition: str) -> "Query.Builder":
            self._where_clauses.append(condition)
            return self

        def order_by(self, column: str) -> "Query.Builder":
            self._order_by = column
            return self

        def limit(self, limit: int) -> "Query.Builder":
            self._limit = limit
            return self

        def offset(self, offset: int) -> "Query.Builder":
            self._offset = offset
            return self

        def build(self) -> "Query":
            if not self._table:
                raise ValueError("Table is required")
            if not self._columns:
                self._columns = ["*"]
            return Query(self)

File: java/src/patterns/creational/builder/Query.java
package patterns.creational.builder;

import java.util.*;

class Query {
    // required
    private final List<String> columns;
    private final String table;

    // optional
    private final List<String> whereClauses;
    private final String orderBy;
    private final Integer limit;
    private final Integer offset;
    // can add more fields as needed

    // Private constructor to prevent direct instantiation
    private Query(Builder builder) {
        this.columns = builder.columns;
        this.table = builder.table;
        this.whereClauses = builder.whereClauses;
        this.orderBy = builder.orderBy;
        this.limit = builder.limit;
        this.offset = builder.offset;
    }

    @Override
    public String toString() {
        return "Query{" +
                "columns=" + columns +
                ", table='" + table + '\'' +
                ", whereClauses=" + whereClauses +
                ", orderBy='" + orderBy + '\'' +
                ", limit=" + limit +
                ", offset=" + offset +
                '}';
    }

    public String toSql() {
        StringBuilder sql = new StringBuilder("SELECT ");
        if(columns.size() == 1 && columns.get(0).equals("*")) {
            sql.append("*");
        } else {
            sql.append(String.join(", ", columns));
        }
        sql.append(" FROM ").append(table);
        if(!whereClauses.isEmpty()) {
            sql.append(" WHERE ").append(String.join(" AND ", whereClauses));
        }
        if(orderBy != null) {
            sql.append(" ORDER BY ").append(orderBy);
        }
        if(limit != null) {
            sql.append(" LIMIT ").append(limit);
        }
        if(offset != null) {
            sql.append(" OFFSET ").append(offset);
        }
        return sql.append(";").toString();
    }

    public static class Builder {
        private final List<String> columns = new ArrayList<>();
        private final String table;
        private List<String> whereClauses = new ArrayList<>();
        private String orderBy;
        private Integer limit;
        private Integer offset;

        public Builder(String table, String... columns) {
            this.table = table;
            this.columns.addAll(Arrays.asList(columns));
        }
        
        // For each method, we return the Builder object itself
        public Builder where(String condition) {
            this.whereClauses.add(condition);
            return this;
        }

        public Builder orderBy(String column) {
            this.orderBy = column;
            return this;
        }

        public Builder limit(Integer limit) {
            this.limit = limit;
            return this;
        }

        public Builder offset(Integer offset) {
            this.offset = offset;
            return this;
        }

        public Query build() {
            if(this.table.isEmpty()) {
                throw new IllegalArgumentException("Table is required");
            }
            if(this.columns.isEmpty()) {
                this.columns.add("*");
            }
            return new Query(this);
        }
    }
}